Server/ReasnAPI/ReasnAPI/Services/EventService.cs

```diff
@@ -1,16 +1,291 @@
-﻿using ReasnAPI.Models.Database;
-
-namespace ReasnAPI.Services {
-    public class EventService (ReasnContext context) {
-        private readonly ReasnContext _context = context;
-
-        /* TODO: Create following functions for this class
-         * create
-         * update
-         * delete
-         * get by ID
-         * get list by filter
-         * get all
-         */
-    }
-}
+﻿using Microsoft.EntityFrameworkCore;
+using ReasnAPI.Exceptions;
+using ReasnAPI.Models.Database;
+using ReasnAPI.Models.DTOs;
+using System.Linq.Expressions;
+using System.Transactions;
+using System.Text.RegularExpressions;
+using ReasnAPI.Mappers;
+
+namespace ReasnAPI.Services;
+public class EventService(ReasnContext context)
+{
+    public EventDto CreateEvent(EventDto eventDto)
+    {
+        using (var scope = new TransactionScope())
+        {
+            var createdTime = DateTime.UtcNow;
+            var newEvent = eventDto.ToEntity();
+            newEvent.CreatedAt = createdTime;
+            newEvent.UpdatedAt = createdTime;
+            newEvent.Slug = CreateSlug(eventDto, createdTime);
+
+            context.Events.Add(newEvent);
+            context.SaveChanges();
+
+            AddTags(eventDto, newEvent);
+            AddParameters(eventDto, newEvent);
+
+            context.SaveChanges();
+            scope.Complete();
+        }
+
+        return eventDto;
+    }
+    private void AddTags(EventDto eventDto, Event newEvent)
+    {
+        if (eventDto.Tags is not null && eventDto.Tags.Any())
+        {
+            var newTags = eventDto.Tags
+                .Select(t => context.Tags.FirstOrDefault(x => x.Name == t.Name) ?? new Tag { Name = t.Name })
+                .ToList();
+
+            context.Tags.AddRange(newTags.Where(t => t.Id == 0)); // Add only new tags to the context
+            newEvent.Tags = newTags;
+        }
+    }
+
+    private void AddParameters(EventDto eventDto, Event newEvent)
+    {
+        if (eventDto.Parameters is not null && eventDto.Parameters.Any())
+        {
+            var newParameters = eventDto.Parameters
+                .Select(p => context.Parameters.FirstOrDefault(x => x.Key == p.Key && x.Value == p.Value) ?? new Parameter { Key = p.Key, Value = p.Value })
+                .ToList();
+
+            context.Parameters.AddRange(newParameters.Where(p => p.Id == 0)); // Add only new parameters to the context
+            newEvent.Parameters = newParameters;
+        }
+    }

```

Comment:
I was thinking of more likely a methods which can be more reusable, so something which gets eventId and list of tags/parameters, so then of course add will not replace event.Parameters but will add to the collection and we can do remove not update to to the opposite. Because there are basically two cases: removing and adding parameters/tags. In event creation we want to add all, in event updating we want to remove all which are different/are no present and add the difference. So then the methods will be reusable and a bit simplier to understand, since basically AddTags/AddParameters is not even having Tags/Parameters in method params, which might be misleading and in Update we don't even update, only Add/Remove, so that's misleading too.

Reply:
So we will then have 4 methods with clear logic instead of 6 methods which are doing kinda the same but the parameters are different and logic is not as we might think when looking at method name